documentation/general/workloads/workload-history.md

```diff
@@ -0,0 +1,44 @@
+
+## Recording installation actions
+
+When a CLI command is run which modifies the installed workloads or updates the manifests, we will write a record to the workload history log.  The workload commands this applies to include `install`, `update`, `uninstall`, `repair`, and `restore`.  Each workload history record will include the following information:
+
+- Date/time command was run
+- Date/time command completed
+- Top-level workload command run (such as install, update, or uninstall)
+- Workloads passed as arguments
+- Contents of rollback file passed in, if any
+- Full command line
+- Success or failure, and error message on failure
+- Workload state before and after command run
+  - Manifest versions (rollback file)
+  - Installed workloads
+
+Each installation record will be stored as a file in a folder specific to the SDK feature band.  For MSI-based installs, this folder will be `%PROGRAMDATA%\dotnet\workloads\{sdk-band}\history\`.  For File-based installs, it will be `{DOTNET ROOT}\metadata\workloads\{sdk-band}\history\`.
+
+## `dotnet workload history`
+
+The `dotnet workload history` command will show the history of workload installation commands.  The output will be similar to the following format:
+
+|ID|Date|Command|Workloads|
+|--|----|-------|---------|
+|3|2023-02-01|Update|android, maui, wasm-tools|
+| |2023-02-01|Unknown|Unknown|
+|2|2023-01-05|Install|maui, wasm-tools|
+|1|2023-01-01|Install|android|
+
+The ID will be an identifier for the record which will start at 1 for the oldest record and increment from there.  Each record includes the workload state before and after the command was run.  The history command will compare the state after each command was run with the state recorded before the next command record.  If there is a difference, then something changed the state that was not recorded, most likely a Visual Studio installation, or an update to the .NET SDK.  In such a case an "unknown" line will be included in the output to represent this difference.
+
+## `dotnet workload update --from-history
+
+A `--from-history` option will be added to the `dotnet workload update` command, which will update to a state from the workload history.  This will require one of two new parameters, `--before` or `--after`, to be specified.  By default the command will update the manifests to the versions specified, as well as install or uninstall workloads to match the workloads that were installed at the specified point.  A `--manifests-only` option will be supported which will update the manifests, but not change the current workloads that are installed (this may result in installing or uninstalling packs to match the new manifest versions).
+
+For example:
+
+`dotnet workload update --from-history --before 3 --manifests-only` - This will roll back to the manifests before operation 3 in the workload history.
```

Comment:
So `before` walks back and selects the first operation that was successful? For example, if you did `--before 8` and 6 and 7 failed, but 5 succeeded, it will select 5? Would `--after 2` select the last known successful state (if 3 and 5 were successful and 4 failed, it would pick 5)?

Reply:
I'd say --before picks the latest successful operation before the id specified, so in your case, it would pick 5, but if 1/2/3/4 were all successful, it would still pick 5. After would pick the first successful one after that, so 3 for your case or 2 if that one was successful. (There has to be some way to say "pick 2 if possible", and after seems more appropriate than before.) Does that sound reasonable to you?

Reply:
No, what I was going for was that `--before n` would give you the state immediately before command n was executed, regardless of whether it succeeded or failed.  If command failed, then ideally it would roll back any changes, and `--before n` and `--after n` would mean the same thing.

Basically each record in the history records the state before and after the command was executed.  The number specified selects which record to use, and whether you specify `--before` or `--after` selects which part of the state in that record to use.  Whether a command was successful or not doesn't affect the logic.

Does this make sense?  If this wasn't obvious from my description, maybe it's also going to be confusing to developers and we should try to find a different way to express these operations.